scripts/update_scpca_checkpoints.py

```diff
@@ -0,0 +1,479 @@
+    metadata_key = f"{dest_prefix}/scpca-meta.json"
+
+    publish_uri = f"s3://{bucket}/{prefix}/{consts.dest_dir}/rad/{run.scpca_library_id}"
+    rad_uri = f"{publish_uri}/{run.scpca_run_id}-rna"
+
+    # set up S3
+    s3 = boto3.resource('s3')
+    s3_bucket = s3.Bucket(bucket)
+
+    ### read output metadata, if present
+    try:
+        results_obj = s3_bucket.Object(results_json)
+        results_meta = json.loads(results_obj.get()['Body'].read().decode('utf-8'))
+    except s3.meta.client.exceptions.NoSuchKey:
+        # object did not exist
+        results_meta = {}
+    scpca_version = results_meta.get("workflow_version")
+    if not scpca_version:
+        scpca_version = consts.scpca_version
+
+    ### Build metadata object
+    metadata = {
+        "run_id": run.scpca_run_id,
+        "library_id": run.scpca_library_id,
+        "sample_id": run.scpca_sample_id ,
+        "project_id": run.scpca_project_id ,
+        "submitter": run.submitter,
+        "technology": run.technology,
+        "seq_unit": run.seq_unit,
+        "feature_barcode_file": run.feature_barcode_file,
+        "feature_barcode_geom": run.feature_barcode_geom,
+        "files_directory": run.files_directory,
+        "slide_serial_number": run.slide_serial_number,
+        "slide_section": run.slide_section,
+        "ref_assembly": consts.ref_assembly,
+        "ref_fasta": consts.ref_fasta,
+        "ref_gtf": consts.ref_gtf,
+        "salmon_splici_index": consts.salmon_splici_index,
+        "salmon_bulk_index": consts.salmon_bulk_index,
+        "t2g_3col_path": consts.t2g_3col_path,
+        "t2g_bulk_path": consts.t2g_bulk_path,
+        "cellranger_index": consts.cellranger_index,
+        "star_index": consts.star_index,
+        "scpca_version": scpca_version,
+        "nextflow_version": consts.nextflow_version,
+        "rad_publish_dir": publish_uri,
+        "rad_dir": rad_uri,
+        "barcode_file": run.barcode_file
+    }
+
+    # print(json.dumps(metadata, indent=2))
+
+    ## copy files from source to destination
+    # list source and destination to see if files exist
+    origin_objs = list(s3_bucket.objects.filter(Prefix = origin_prefix))
+    dest_objs = list(s3_bucket.objects.filter(Prefix = dest_prefix))
+    if len(origin_objs) == 0:
+        print(f"No files for {run.scpca_run_id}")
+    elif len(dest_objs) > 0 and not overwrite:
+        print(f"Files present at destination for {run.scpca_run_id} -- skipping")
+    else:
+        print(f"Copying files for {run.scpca_run_id}")
+        ### S3 copying using awscli
+        sync_command = [
+            "aws", "s3", "sync",
+            f"s3://{bucket}/{origin_prefix}",
+            rad_uri
+        ]
+        subprocess.run(sync_command)
+        ### write json object
+        s3_bucket.put_object(
+            Key = metadata_key,
+            Body = json.dumps(metadata, indent=2)
+        )
```

Comment:
I feel like you could create one function that does builds the metadata and then does the copying. After building the uris and prefixes that are unique to each technology you could have one function that takes as input those uris and the `run` object. That function would create the metadata and do the copying so you wouldn't have the same code repeated.

Reply:
Yes, I definitely thought about that, but the metadata we need varies slightly among the data types. So I could do _most_ of the values with a function, but then I would need to add the specific values somehow... This way also had the advantage that I could check that I got everything by comparing to an existing file for that checkpoint type relatively easily.

With the writing of files, there were some kind of subtle things with reusing the s3 bucket object that I was trying to avoid.

All of these are surmountable, but I'm really hoping that for a script that will  hopefully be run only once, it won't matter. (subtext: I don't want to work on this more than I need to!)

Reply:
I now see the `rad_publish_dir` at the bottom of the metadata file that changes for each method, which makes sense why you did it individually.